keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+to gain control of the scheduler, get it to execute API calls, and then
+retrieve the result. This seems like a low risk.
+
+Using a single NodeResourceCapacity object per node is simple and convenient
+also for admins, but might run into size limitations. This should not be a
+problem in practice because the number of drivers per node is expected to be
+low and the capacity information from each should be small. Limits for the
+JSON fragment can be introduced to prevent excessive sizes.
+
+## Design Details
+
+### Decoding claim parameters
+
+Extracting the numeric parameters can be done without knowing the CRD:
+
+- Look up the resource based on group and kind with a REST mapper.
+
+- Retrieve the object with a `dynamic.Client` as an `unstructured.Unstructured`
+  instance (basically a nested map).
+
+- Retrieve the sub-map via the JSON path.
+
+- Re-encode that part as JSON with the normal `encode/json` package.  In the
+  example above, this will include `kind` and `apiVersion:
+  counter.dra.config.k8s.io` because those are the defaults and only allowed
+  values.
+
+- Decode into the numeric parameter with the apimachinery JSON decoder,
+  ignoring unknown fields like `env` in this example.
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables numeric parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via numeric parameters, then NumericParameters must list all claim
+    // parameter types that are allowed for claims.
+    NumericParameters []NumericParameterType
+}
+
+type NumericParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// FieldPath is a dot separarated JSON path that defines where in the
+	// parameter object the numeric parameters are embedded.
+	FieldPath string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResourceCapacity
+
+For each node, a NodeResourceCapacity objects gets created:
+
+```
+// NodeResourceCapacity gets published by kubelet. Its name
+// matches the name of the node and the node is the owner.
+//
+// Capacity may get added, but should not get removed because
+// it would make scheduling decisions based on the old capacity
+// invalid.
+type NodeResourceCapacity struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Resources contains information about the capacity
+	// reported by each DRA driver.
+	//
+	// +optional
+	Resources []DriverResources
```

Comment:
This object combines the resource information of all drivers on the node. This was based on:
- convenience for users (`kubectl get noderesourcecapacity/<node name>`)
- ease of implementation (no need to track objects because of the deterministic name), simple implementation in the node authorizer

But overall size is a concern here. I think I now prefer to split this up:
```
type NodeResourceCapacity struct {
...
NodeName string
DriverName string
ResourceInstance runtime.RawExtension
}
```

We can still make this convenient for the user by allowing to filter by nodename and drivername field.

The kubelet needs to use a similar filter in its watch to only receive objects that it owns. I need to check how the node authorizer can restrict create and read access.

cc @bart0sh @byako

Reply:
One positive side effect of the revised API is that E2E testing is easier to parallelize: publishing NodeResourceCapacity for one driver is independent from that for other drivers.

Reply:
So, I think we should *probably* define this structure rather than leaving it as a RawExtension. Since we are defining numerical models anyway, why wouldn't we define the capacity structure?

In fact, perhaps I am misunderstanding something about the proposal. As I understand it, the NumericParameterTypes defined in the resource class tell us how to extract the relevant parameters from the claim parameters (putting those into a K8s, driver-neutral list of numerical requests). But if the capacity structure is opaque, how to do we do the fit evaluation? What am I missing? Perhaps it's that the individual numerical requests have some sort of name or path that associates them to a field in the opaque capacity structure? If that's the case, we may be able to get away with that at first, but I suspect the name mapping may not be enough; we may need another CEL transformation to represent the allocation process (if it's a more complex allocation).

Another option would be to combine the steps, where instead of the resource class including a list of "extraction" CEL expressions, it instead includes a set of "allocation" expressions. So, the inputs to those expressions would be the node capacity and the claim parameters. The output would be the new node capacity after that allocation. We would also probably need a validation expression to identify if that is a valid allocation (for example, a counter can't go negative).

I think that could work. The only issue is I am not sure it provides *visibility* for the scheduler and/or CA into the reasons allocation may be failing, so that they can make smarter choices. It sort of forces a random walk of the search space when trying to schedule or find a valid new node type to create. I'd like to be able to provide a model that can be interpreted. But perhaps that's a later step.

Reply:
> So, I think we should probably define this structure rather than leaving it as a RawExtension. Since we are defining numerical models anyway, why wouldn't we define the capacity structure?

Several reasons:
- There's more then one capacity type (one per numeric model). Instead of a single field we would need a struct which must have one (and only one) of several different alternatives.
- The capacity type needs to be registered in the apiserver. kubelet needs to know about them, too. With the current proposal, adding a new numeric model can be done without changing the apiserver and kubelet. This would be important for someone building their own scheduler with some additional, non-core numeric model.

Reply:
Coming back to the question around object size vs. number of objects - let's go for a compromise and use this:

```
type NodeResourceCapacity struct {
...
NodeName string
DriverName string
Instances []struct {
   InstanceID string
   Capacity runtime.RawExtension
}
```

Most likely, all instances of a single driver will fit into one object, so we end up with a handful of objects per node (one for each driver). The API would be defined from the start such that it's allowed to have multiple objects per driver. If kubelet detects that a single objects gets too large, it can create multiple.

But we don't need to implement that in 1.30 because this overflow detection and recovery would be complicated. Always creating a single object should be good enough for alpha.

@bart0sh: I'll revise the prototype accordingly on Monday. Sorry for the back and forth...

Reply:
@johnbelamaric: let's continue the discussion around `RawExtension` here. I listed the main advantages of that approach in https://github.com/kubernetes/enhancements/pull/4384/files#r1467587340.

From a practical perspective, using `RawExtension` enables delayed updates of kubelet: kubelet can remain at v1.30 while control plane and DRA drivers already get updated to a future version with a different numeric model. I believe it's not unusual to skip kubelet updates because it's more work, more disruptive, and harder to roll back than control plan updates.

If we do something like you proposed in https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1912886646 where `NodeResourceCapacity` depends on the capacity types of the numeric models, where would you define those types?

If in the `resource.k8s.io` API, then kubelet needs to link against `k/k/pkg/apis` to do the conversion and validation. If in the `k8s.io/dynamic-resource-allocation` API (my current proposal), then the `resource.k8s.io` API depends on that. Both seem doable to me, but neither is ideal.

Reply:
The drawback of `RawExtension` is delayed validation: kubelet and apiserver accept raw data, only the final recipient , the scheduler, parses and checks it.

For a user API that is a pretty big no-no, but here we are talking about validating data provided by a DRA driver. I think it's okay to put a bit more trust into DRA drivers (it is a privileged component on the node) and their developers (there are less of them than end-users of Kubernetes; if they make a mistake. it is limited to the usage of their DRA driver).

Reply:
Side note: instead of `RawExtension`, let's use [`Unknown`](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#Unknown). That makes `kind` and `apiVersion` explicit at the core API level, which is better than implicitly requiring it inside the raw data.

Reply:
Using `Unknown` didn't work out, encoding to/from JSON ignores the `kind` and `apiVersion`. But lifting those fields up still makes sense and can be dong by adding `Type metav1.TypeMeta`.

Reply:
@johnbelamaric: let's use this thread for "strongly typed" vs. "pass from driver to scheduler" (= `RawExtension`).

You wrote:
> Putting the numerical models behind raw extension doesn't make them NOT part of the K8s API - it just makes them an unmanaged extension to the API that lacks the consistent versioning and standard deprecation controls and policies.

No, that's not correct. They become APIs under k8s.io/dynamic-resource-allocation/apis, with an `OWNERS` file that ensures that api reviews are done for them. This is the same approach that is taken for other Kubernetes APIs like [the kubelet config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/OWNERS). The normal Kubernetes rules and processes for managing APIs also apply to the numeric models.

Let me repeat the benefits of the current design:
- supports version skew (kubelet older than control plane)
- allows building custom schedulers with custom numeric types without having to modify Kubernetes first (not a priority, but might become important for some users)

cc @liggitt

Reply:
I didn't read this thread in detail yet (I will get to it soon), but I wanted to throw in my support for keeping `RawExtension`s here. As pointed out by @pohly and @johnbelamaric, it's DRA driver implementors (not end-users) that will be defining the types that map to these RawExtensions. So while we do lose the ability to do proper validation at the time one of these CRs is created, the added flexibility is justified by the fact that only a handful of experts will be populating these fields in practice.

Reply:
Independent of that, I also had a general question about how this type relates back to the `Capacity` type discussed in the original example (i.e. the one with just a single `Count` field).

Assuming this was the only numeric parameter supported by a driver, would an instance of this `Capacity` type be what gets set assigned to `Instances[0].Capacity` below?: 
```
type NodeResourceCapacity struct {
...
NodeName string
DriverName string
Instances []struct {
   InstanceID string
   Capacity runtime.RawExtension
}
```

Reply:
@klueska: correct.

Reply:
replying specifically to all the threads about RawExtension and whether the core API types should be defined with schemas or opaque and unvalidated - threads like:
* https://github.com/kubernetes/enhancements/pull/4384#discussion_r1467101423
* https://github.com/kubernetes/enhancements/pull/4384#discussion_r1468500718
* https://github.com/kubernetes/enhancements/pull/4384#discussion_r1469580780
* https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1913108043
* https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1913255154

---

From https://github.com/kubernetes/enhancements/pull/4384#discussion_r1467587340:

> There's more then one capacity type (one per numeric model). Instead of a single field we would need a struct which must have one (and only one) of several different alternatives.

That sounds exactly like the [union type](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#Unions) pattern we use everywhere in the API where we have multiple mutually-exclusive options with their own schemas, validation, etc.:

> API objects often are [union](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#Unions) object containing the following:
>
> 1. One or more fields identifying the Type specific to API object (aka the discriminator).
> 2. A set of N fields, only one of which should be set at any given time - effectively a union.
>
> Controllers operating on the API type often allocate resources based on the Type and/or some additional data provided by user.

That allows for strongly-typed and validated data while allowing for future expansion.

---

From https://github.com/kubernetes/enhancements/pull/4384#discussion_r1467587340:

> With the current proposal, adding a new numeric model can be done without changing the apiserver and kubelet. This would be important for someone building their own scheduler with some additional, non-core numeric model.

Wanting to do that for the kubelet seems good... thinking about how to make kubelet a lossless pipe to be able to accommodate future additions on older kubelets is a good thing to think about, but I don't think that requires using RawExtension in the type definition itself. If the bulk of this API is just the accumulated capacity information, the kubelet could use something like the dynamic client to pass accumulated capacity info opaquely to the API server with [strict field validation](https://github.com/kubernetes/kubernetes/blob/release-1.29/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L569-L586).

For the API server, though, if the point of these is to custom DRA things into well-understand things the scheduler can interpret, the API server's _job_ is to make sure persisted stuff is valid, so it's confusing to want to change or define new well-understood numeric models _without_ changing the API server and scheduler.

---

From https://github.com/kubernetes/enhancements/pull/4384#discussion_r1468501870:
> The drawback of RawExtension is delayed validation: kubelet and apiserver accept raw data, only the final recipient, the scheduler, parses and checks it.

There's no precedent for accepting and persisting arbitrary structured data without validation into core REST APIs which is required to have a consistent structure and interpretation by core Kubernetes components. We work _really_ hard to move as much validation as early as possible in the pipeline since propagating errors back to the user gets harder the more async API write layers removed we are from the thing that persisted something problematic. We also want to do the validation once at write time centrally in the API server, not repeatedly at read time distributed in multiple reading components.

---

From https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1912886646:

> NodeResourceCapacity is no longer a RawExtension but instead a collection of well-defined types with commonly undertood semantics. Individual entries have names and each has an associated numerical model type.

That roughly matches my expectation. I'd probably add a type discriminator field as well. We could start with the types we understand today, and if we need new models in the future (as I expect we will), the union type would grow new options.

---

From https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1913108043:

> So in other words, individual numeric models become part of the Kubernetes API. What is the advantage of that?

API documentation, API server validation, schema validation, clear stability guarantees, etc

> Your initial reaction to seeing RawExtension probably was "this is opaque, that's bad". But it is not opaque where it matters (scheduler, autoscaler) because those components who care about the capacity can decode it using Kubernetes-defined API types.

I don't understand this point, couldn't the same be said about most APIs that are only read by a small number of components? It moves failure modes to the point of reading and makes all of those components responsible for validation, which doesn't seem good.

> The downsides are the higher bar that we have to clear with designing these numeric models.

I think the bar is the same regardless of where we define the type.

> If we get it wrong, the resource.k8s.io API is stuck with supporting failed experiments because e.g. ResourceCapacity.Block will forever be associated with this "block" model.

We've dropped alpha fields before, so if an experiment fails before alpha → beta you can tombstone a field name and not be stuck with it (though you can't reuse the same field name for a different schema later).

---

My summary:

* Fields that are written to the API server in built-in APIs and are read and interpreted by core components like scheduler should be understood and validated by the API server, and should not use RawExtension.
* Keeping type safety and validation concerns with a single actor once at write time is greatly preferred over multiple actors at read time.
* Allowing for future expansion / future models is a good concern. A union type allows for that.
* Allowing for old kubelets to opaquely create these objects is a good concern. I think there are ways to address that (like using the dynamic client) without giving up type information and validation at the API and API server layer.

Reply:
> That sounds exactly like the [union type](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#Unions) pattern we use everywhere in the API

Yes, it is. Let's spell it out. This would be the type definition in `resource.k8s.io/v1alpha2`:
```
import (
    // Some numeric models...
    counterv1alpha1 "k8s.io/dynamic-resource-allocation/apis/counter/v1alpha1"
    modeswitchv1alpha1 "k8s.io/dynamic-resource-allocation/apis/modeswitch/v1alpha1"
)

type NodeResourceInstance struct {
	ID string
	
        Model NodeResourceModels
}

// Must have exactly one field set.
type NodeResourceModels {
    Counter *counterv1alpha1.Capacity
    ModeSwitch *modeswitchv1alpha1.Capacity
}
```

Note that this tries to keep the semantic models separate from each other and the common `resource.k8s.io` API. That works for the public types. It does not work for the internal types unless we also export them in `k8s.io/dynamic-resource-allocation`. So perhaps this should be under `k8s.io/api` (external) and `k/k/pkg/apis` after all (but see below)?

We need to do the same for the allocation result. Currently that is stored in `Data`:
```
type ResourceHandle struct {
	Data string
}
```

We need to extend that with pointers:
```
type ResourceHandle struct {
	Data ResourceHandleData
}

// Warning, API break between v1alpha2 and v1alpha3 (assuming we switch to that).
type ResourceHandleData {
        Driver *string
	Counter *counterv1alpha1.AllocationResult
	ModeSwitch *modeswitchv1alpha1.AllocationResult
}
```

> > With the current proposal, adding a new numeric model can be done without changing the apiserver and kubelet. [...]
>
> Wanting to do that for the kubelet seems good... thinking about how to make kubelet a lossless pipe to be able to accommodate future additions on older kubelets is a good thing to think about, but I don't think that requires using RawExtension in the type definition itself. If the bulk of this API is just the accumulated capacity information, the kubelet could use something like the dynamic client to pass accumulated capacity info opaquely to the API server with [strict field validation](https://github.com/kubernetes/kubernetes/blob/release-1.29/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L569-L586).

So kubelet stores a `NodeResourceCapacity` as `unstructured.Unstructured`, receives some data from the DRA driver for `counter`, parses it into `map[string]any` and stores that in the `unstructured.Unstructured` map, then finally submits to the apiserver. I can see how that'll work.

Well, not quite. kubelet needs to pick a version for the `NodeResourceCapacity` when setting up a watch, with a field filter for the node name. We want to cache existing objects of that resource to detect when the data is stale.
Suppose that's `v1alpha3` (the version known to kubelet). That means all DRA drivers on the node must provide data that matches `counterv1alpha1.AllocationResult` or `modeswitchv1alpha1.AllocationResult`, otherwise kubelet cannot write back.

`NodeResourceCapacity` objects are per-driver. I suppose it would work if the driver told kubelet "store this data in `counter` using NodeResourceCapacity v1beta1". This works as long `v1alpha3` and `v1beta1` are "similar enough" - i.e. the rest of the fields don't change. Could work for `NodeResourceCapacity`. We would need to give up reusing existing objects: if kubelet is seeing a revision that it hasn't written itself, it has to overwrite, because it cannot compare `NodeResourceModels` (potential version mismatch). 

I don't see how to make it work efficiently for `ResourceHandleData`. Here kubelet needs to receive `ResourceClaim` as strongly-typed objects, with `v1alpha2`. To get the full data for the driver, it would need to do one GET with the driver-specific version. Perhaps okay performance-wise?

Suppose we give up "kubelet as lossless pipe". It then needs access to the internal type definitions, otherwise it cannot convert between `v1alpha1` (sent by an old driver) and a newer `v1alpha2` (used by kubelet). I don't think it should link against `k/k/apis`, so let's keep the types in `k8s.io/dynamic-resource-allocation` with just the schema registration exported, but not the internal types?

> API documentation, API server validation, schema validation, clear stability guarantees, etc

- API documentation: these types are for DRA driver developers. I think they'll be more concerned about good godoc documentation and less about the generated API documentation.
- API server validation: ack. I agree that this is the main weakness of the `RawExtension` proposal.
- schema validation: the types of the semantic model would also be available for registration in a scheme, so this would be covered (if I understand the difference to "API server validation" correctly)
- clear stability guarantees: this is covered by an OWNERS file that makes the semantic model types part of the Kubernetes API (same as kubelet config)

Reply:
This format (GH conversation threads) is really breaking down.  Patrick, you're making jumps that I don't follow.

What I read in Jordans remarks was that it would be better to define a small set of strongly typed and managed parameter semantics, which can be extended in the "usual" kubernetes ways.

I don't know how to track th erest of your response - you have some context I don't. :(

Reply:
Your take away from Jordan's comment is correct. The other key take away is that allowing kubelet to be older than the control plane and the driver is useful.

This is where strong typing causes a problem. Jordan outlined a potential solution, but it has gaps that I tried to explain in my response, hoping that Jordan might offer some more advice. I have thought about it some more and now see a path forward that combines "strong typing" with "kubelet a lossless pipe".

But let's not discuss that path further now. As a first step we can ignore "kubelet a lossless pipe" and add it later. It won't make a difference for 1.30 (only one version) and an API break in 1.31 would be okay (alpha APIs in 1.30). A solution in 1.30 would be better, but that can be discussed together with the implementation. IMHO it doesn't need to be in the KEP from the start.